File: impls/python/milpa/identity.py
"""Content-addressed identity — the canonical content Merkle DAG of a source tree.

milpa's *identity* of a dep is the ``dag-sha256:`` digest of the canonical content
Merkle DAG (spec/identity.md §1.8) of its source tree. The hash is
transport-independent, provenance-independent, and recomputable from the bytes on
disk alone.

Computation (identity.md §1.8, epoch 2 — the forever format):
    ``compute_content_hash(path)`` walks the on-disk tree into a buffered
    ``list[MaterializedEntry]`` (the **canonical disk walk**, ``enumerate_local_entries``)
    and feeds it to the pure Merkle-DAG builder
    (``milpa.dag_identity.compute_dag_identity``). The disk walk is the single
    source of truth for turning real on-disk bytes + POSIX mode bits into the
    materialize seam; every transport that lands a tree on disk (git, tarball,
    local) inherits this identity uniformly, and CAS verify re-hashes the stored
    tree through the same walk.

    Cross-cutting content rules (shared with the per-transport object-store seams):
      - `.git` excluded at any depth (§1.4 / §1.8.6).
      - symlinks hashed by their UTF-8 target string, never followed (§1.5).
      - raw bytes, no line-ending normalization (§1.6).
      - the executable bit is part of identity (epoch 2, §1.8.2 — mode-byte 0x01).
      - non-UTF-8 relpaths / symlink targets are coded errors (§1.5).

Identity string (identity.md §2.1):
    dag-sha256:<64-lowercase-hex-chars>

    The scheme prefix is ``dag-sha256`` and now names the *actual* canonical
    content Merkle DAG of §1.8 (the interim epoch-1 flat byte stream is retired).
    A stale ``sha256:`` prefix is rejected as ID-UNSUPPORTED-ALGORITHM at parse
    time — re-lock with `milpa fetch`.

parse_identity five ordered checks (identity.md §2.2):
    1. Must be a string              → ID-NOT-A-STRING
    2. Must contain ':'              → ID-NO-ALGORITHM-PREFIX
    3. Algorithm in CANONICAL_SCHEMES → ID-UNSUPPORTED-ALGORITHM
       (sha256: and all other prefixes are rejected; only dag-sha256: is canonical)
    4. Digest length correct         → ID-WRONG-DIGEST-LENGTH
    5. Digest all lowercase hex      → ID-NON-HEX-DIGEST
"""

from __future__ import annotations

import os
from pathlib import Path
from typing import Any

from milpa.dag_identity import (
    MODE_EXECUTABLE,
    MODE_REGULAR,
    MODE_SYMLINK,
    MaterializedEntry,
    compute_dag_identity,
)
from milpa.errors import (
    ID_NO_ALGORITHM_PREFIX,
    ID_NON_HEX_DIGEST,
    ID_NON_UTF8_RELPATH,
    ID_NON_UTF8_SYMLINK_TARGET,
    ID_NOT_A_STRING,
    ID_UNSUPPORTED_ALGORITHM,
    ID_WRONG_DIGEST_LENGTH,
    MilpaError,
)

# ---------------------------------------------------------------------------
# Algorithm table (identity.md §2.2 two-tier scheme classification)
# ---------------------------------------------------------------------------

#: Canonical scheme set — the ONLY accepted prefixes in identity strings.
#: Stale sha256: and all other prefixes are rejected as ID-UNSUPPORTED-ALGORITHM.
#: Adding a future canonical scheme: add it here and to _DIGEST_HEX_LEN below.
SUPPORTED_ALGORITHMS: frozenset[str] = frozenset({"dag-sha256"})

#: Expected hex-character length of the digest part for each canonical algorithm.
_DIGEST_HEX_LEN: dict[str, int] = {"dag-sha256": 64}


# ---------------------------------------------------------------------------
# parse_identity — 5 ordered checks (identity.md §2.2)
# ---------------------------------------------------------------------------


def split_identity_scheme(s: Any) -> tuple[str, str]:
    """Split an identity string into its raw ``(algorithm, digest)`` halves.

    Applies only the first two ordered checks from identity.md §2.2 — "must be
    a string" and "must contain a ``:`` separator" — WITHOUT enforcing the
    canonical-scheme allowlist (checks 3-5, ``SUPPORTED_ALGORITHMS``). This is
    the shared validation seam for callers that only need the raw scheme
    split and must not silently no-op on malformed input — currently
    ``parse_identity`` itself, and ``entry_trust.build_entry_subject``, which
    extracts a hex digest from a ``content_hash`` without needing (or
    wanting) to couple itself to ``SUPPORTED_ALGORITHMS``.

    Raises:
        MilpaError(ID_NOT_A_STRING)        — input is not a str
        MilpaError(ID_NO_ALGORITHM_PREFIX) — no ':' separator
    """
    if not isinstance(s, str):
        raise MilpaError(
            ID_NOT_A_STRING,
            f"identity must be a string, got {type(s).__name__!r}",
            got_type=type(s).__name__,
        )
    if ":" not in s:
        raise MilpaError(
            ID_NO_ALGORITHM_PREFIX,
            f"identity {s!r} is missing the algorithm prefix; "
            f"expected '<algorithm>:<digest>' (e.g. 'dag-sha256:abc...')",
            identity=s,
        )
    algorithm, _, digest = s.partition(":")
    return algorithm, digest


def parse_identity(s: Any) -> str:
    """Validate a milpa identity string (two-tier scheme check).

    Applies the five ordered checks from identity.md §2.2 in order.
    Returns the input string unchanged when valid (the canonical form IS the
    input itself).

    Only dag-sha256: is accepted (CANONICAL_SCHEMES = SUPPORTED_ALGORITHMS).
    sha256: and all other prefixes → ID-UNSUPPORTED-ALGORITHM (no legacy tier;
    re-lock with `milpa fetch`).

    Raises:
        MilpaError(ID_NOT_A_STRING)            — input is not a str
        MilpaError(ID_NO_ALGORITHM_PREFIX)     — no ':' separator
        MilpaError(ID_UNSUPPORTED_ALGORITHM)   — algorithm not in SUPPORTED_ALGORITHMS
                                                 (includes stale sha256: identities)
        MilpaError(ID_WRONG_DIGEST_LENGTH)     — digest length wrong for algorithm
        MilpaError(ID_NON_HEX_DIGEST)         — digest contains non-lowercase-hex chars
    """
    # Checks 1-2: must be a string, must contain ':' separator (identity.md
    # §2.2 rules 1-2) — delegated to the shared split_identity_scheme seam.
    algorithm, digest = split_identity_scheme(s)

    # Check 3: algorithm must be in the canonical scheme set (identity.md §2.2 rule 3).
    # Two-tier: CANONICAL_SCHEMES = {dag-sha256}; everything else (including the
    # legacy sha256: prefix) is rejected with no compatibility path — re-lock.
    if algorithm not in SUPPORTED_ALGORITHMS:
        allowed = ", ".join(sorted(SUPPORTED_ALGORITHMS))
        hint = (
            " (stale epoch-1 hash — re-lock with `milpa fetch`)"
            if algorithm == "sha256"
            else ""
        )
        raise MilpaError(
            ID_UNSUPPORTED_ALGORITHM,
            f"identity {s!r} uses unsupported algorithm {algorithm!r}"
            f"{hint} (supported: {allowed})",
            algorithm=algorithm,
            identity=s,
        )

    expected_len = _DIGEST_HEX_LEN[algorithm]

    # Check 4: digest length must be correct (identity.md §2.2 rule 4)
    if len(digest) != expected_len:
        raise MilpaError(
            ID_WRONG_DIGEST_LENGTH,
            f"identity {s!r}: {algorithm} digest must be exactly "
            f"{expected_len} hex characters, got {len(digest)}",
            algorithm=algorithm,
            expected=expected_len,
            got=len(digest),
            identity=s,
        )

    # Check 5: digest must be lowercase hex only (identity.md §2.2 rule 5)
    if not all(c in "0123456789abcdef" for c in digest):
        raise MilpaError(
            ID_NON_HEX_DIGEST,
            f"identity {s!r}: digest must be lowercase hex characters (0-9, a-f)",
            identity=s,
        )

    return s


# ---------------------------------------------------------------------------
# compute_content_hash (identity.md §1.8) — the production epoch-2 emitter
# ---------------------------------------------------------------------------


def compute_content_hash(path: Path) -> str:
    """Compute the canonical content Merkle-DAG identity of the tree at ``path``.

    Returns the identity string ``dag-sha256:<64-hex>`` (identity.md §2.1). The
    digest is the root of the canonical content Merkle DAG of §1.8: the on-disk
    tree is walked into a buffered ``list[MaterializedEntry]`` by
    ``enumerate_local_entries`` and folded by ``compute_dag_identity``.

    Same source bytes always produce the same hash regardless of how the tree
    was obtained (git, tarball, OCI, local copy) — identity is transport- and
    provenance-independent. The STEP-1 invariant (a git-materialized on-disk tree
    hashed here equals the git object-store enumeration) is what lets every
    transport, plus CAS verify, share this one identity site.

    Raises:
        MilpaError(ID_NON_UTF8_RELPATH)        — an on-disk name is not valid UTF-8.
        MilpaError(ID_NON_UTF8_SYMLINK_TARGET) — a symlink's target is not valid UTF-8.
        MilpaError(ID_NAME_TOO_LONG)           — a path component exceeds the
                                                  4096-byte leaf-name ceiling (§1.8.8).
    """
    return compute_dag_identity(enumerate_local_entries(Path(path)))


# ---------------------------------------------------------------------------
# enumerate_local_entries — the canonical on-disk walk (the local materialize seam)
# ---------------------------------------------------------------------------


def enumerate_local_entries(root: Path) -> list[MaterializedEntry]:
    """Walk an on-disk source directory into a buffered ``list[MaterializedEntry]``
    (spec §1.8.4) — the canonical disk walk that feeds the epoch-2 DAG builder.

    This is the single source of truth for turning real on-disk bytes + POSIX
    mode bits into the materialize seam. It is the universal identity walk:
    ``compute_content_hash`` uses it, and so does the local transport (the
    "local materialize seam" is exactly this walk applied to a local source dir,
    re-exported from ``milpa.fetchers.local`` for the per-transport sibling
    narrative). It is the local sibling of the object-store seams
    ``enumerate_git_entries`` / ``enumerate_tarball_entries``: a directory laid
    out from a git tree or a tarball reproduces the same ``dag-sha256:`` (spec §1.1).

    Mode mapping (spec §1.8.2.1):
      * regular file with any POSIX execute bit (``st_mode & 0o111``) → ``0x01``.
      * regular file with no execute bit → ``0x00``.
      * symlink → ``0x80``; content is the link-target string bytes (``readlink``),
        the link is not followed.
      * directories synthesise subtrees in the builder; a ``.git`` directory is
        skipped (not recursed) per §1.8.6.

    Args:
        root: Absolute path to the source directory to materialize. Symlinks within
              the tree are recorded as ``0x80`` leaves; the tree under ``root`` is
              walked without following symlinked directories.

    Returns:
        Buffered ``list[MaterializedEntry]`` (blobs + symlinks), POSIX relpaths.

    Raises:
        MilpaError(ID_NON_UTF8_RELPATH)        — an on-disk name is not valid UTF-8.
        MilpaError(ID_NON_UTF8_SYMLINK_TARGET) — a symlink target is not valid UTF-8.
    """
    root = Path(root)
    entries: list[MaterializedEntry] = []

    def _walk(dir_abs: str, prefix: str) -> None:
        with os.scandir(dir_abs) as it:
            children = sorted(it, key=lambda e: e.name)
        for entry in children:
            # §1.8.6 (inherits §1.4): exclude `.git` at any depth (skip recursing).
            if entry.name == ".git":
                continue
            relpath = _check_local_utf8(
                f"{prefix}{entry.name}",
                ID_NON_UTF8_RELPATH,
                f"local source path component {entry.name!r} is not valid UTF-8",
            )
            if entry.is_symlink():
                target = _check_local_utf8(
                    os.readlink(entry.path),
                    ID_NON_UTF8_SYMLINK_TARGET,
                    f"local symlink {relpath!r} target is not valid UTF-8",
                )
                entries.append(
                    MaterializedEntry(relpath, MODE_SYMLINK, target.encode("utf-8"))
                )
            elif entry.is_dir(follow_symlinks=False):
                _walk(entry.path, relpath + "/")
            elif entry.is_file(follow_symlinks=False):
                content = Path(entry.path).read_bytes()
                st_mode = entry.stat(follow_symlinks=False).st_mode
                mode_byte = MODE_EXECUTABLE if (st_mode & 0o111) else MODE_REGULAR
                entries.append(MaterializedEntry(relpath, mode_byte, content))
            # device nodes / FIFOs: silently skipped (never legitimate source).

    _walk(str(root), "")
    return entries


def _check_local_utf8(s: str, slug: str, message: str) -> str:
    """Return ``s`` if it round-trips through UTF-8, else raise ``MilpaError(slug)``.

    ``os.scandir`` / ``os.readlink`` decode names with ``surrogateescape``, so a
    non-UTF-8 byte sequence survives as lone surrogates that fail to re-encode.
    """
    try:
        s.encode("utf-8")
    except UnicodeEncodeError as exc:
        raise MilpaError(
            slug, message, value=s.encode("utf-8", "backslashreplace").decode()
        ) from exc
    return s

File: impls/rust/crates/milpa-core/src/identity.rs
//! Content-addressed identity (RFC §4.1; identity.md). The byte-exact content
//! hash algorithm lives ONLY here (SSOT). Two-table algorithm-agility dispatch
//! (`SUPPORTED_ALGORITHMS` + digest lengths, not a hardcoded `== "sha256"`) so
//! future multihash is a one-file change (identity.md §2.3).
//!
//! S4: the canonical byte-stream walk + sha256 digest, plus the identity-string
//! validator (`parse_identity`). The byte layout is normative — see identity.md
//! §1.2: per entry `<relpath> 0x00 <mode-marker> 0x00 <content> 0x00`, entries
//! in ascending raw-byte relpath order, fed to a single live sha256 accumulator.

use std::os::unix::ffi::OsStrExt;
use std::path::Path;

use crate::dag_identity::{
    compute_dag_identity, MaterializedEntry, MODE_EXECUTABLE, MODE_REGULAR, MODE_SYMLINK,
};
use crate::error::CoreError;

/// Hash algorithms milpa understands, with their hex digest length. New
/// algorithms are added here and nowhere else.
pub const SUPPORTED_ALGORITHMS: &[(&str, usize)] = &[("dag-sha256", 64)];

/// Non-catalog sentinel for filesystem I/O failures encountered while walking a
/// tree to hash it. identity.md §5 catalogs no code for this: the Python
/// reference simply propagates a raw `OSError` (`p.read_bytes()` / `p.stat()` /
/// `os.readlink()` are unguarded). It is unreachable on a freshly-materialized
/// tree and not fixture-expressible, so it is deliberately kept OUT of
/// `CoreError::all_codes()` — it is an infrastructure failure, not a catalog
/// condition. The `MILPA-INTERNAL-` prefix marks it as non-catalog. Shared with
/// `store.rs`, whose mkdir/rename/symlink I/O the spec likewise leaves uncoded.
pub(crate) const INTERNAL_IO: &str = "MILPA-INTERNAL-IO";

/// Split an identity string into its raw `(algorithm, digest)` halves on the
/// first `':'` (identity.md §2.2 rule 2) — WITHOUT enforcing the
/// canonical-scheme allowlist (`SUPPORTED_ALGORITHMS`; rules 3-5). This is the
/// shared validation seam for callers that only need the raw scheme split and
/// must not silently no-op on malformed input — currently `parse_identity`
/// itself, and `entry_trust::build_entry_subject`, which extracts a hex digest
/// from a `content_hash` without needing (or wanting) to couple itself to
/// `SUPPORTED_ALGORITHMS`.
///
/// Raises `ID-NO-ALGORITHM-PREFIX` when there is no `':'` separator at all —
/// the failure mode a bare `split_once(':').unwrap_or(...)` would otherwise
/// swallow silently.
pub fn split_identity_scheme(s: &str) -> Result<(&str, &str), CoreError> {
    s.split_once(':').ok_or_else(|| {
        CoreError::Identity(
            "ID-NO-ALGORITHM-PREFIX",
            format!(
                "identity {s:?} is missing the algorithm prefix; \
                 expected '<algorithm>:<digest>' (e.g. 'sha256:abc...')"
            ),
        )
    })
}

/// Validate a multihash-encoded identity string against the `<algorithm>:<digest>`
/// grammar (identity.md §2.2). Returns the `(algorithm, digest)` split on success.
///
/// The Python reference's `ID-NOT-A-STRING` check is unreachable here: the input
/// is statically a `&str`, so the type system enforces it. The remaining four
/// codes mirror `identity.py:parse_identity`.
pub fn parse_identity(s: &str) -> Result<(&str, &str), CoreError> {
    let (algorithm, digest) = split_identity_scheme(s)?;
    let Some((_, expected_len)) = SUPPORTED_ALGORITHMS.iter().find(|(a, _)| *a == algorithm) else {
        let allowed: Vec<&str> = SUPPORTED_ALGORITHMS.iter().map(|(a, _)| *a).collect();
        return Err(CoreError::Identity(
            "ID-UNSUPPORTED-ALGORITHM",
            format!(
                "identity {s:?} uses unsupported algorithm {algorithm:?} (supported: {})",
                allowed.join(", ")
            ),
        ));
    };
    if digest.len() != *expected_len {
        return Err(CoreError::Identity(
            "ID-WRONG-DIGEST-LENGTH",
            format!(
                "identity {s:?}: {algorithm} digest must be exactly {expected_len} \
                 hex characters, got {}",
                digest.len()
            ),
        ));
    }
    if !digest
        .bytes()
        .all(|b| b.is_ascii_digit() || (b'a'..=b'f').contains(&b))
    {
        return Err(CoreError::Identity(
            "ID-NON-HEX-DIGEST",
            format!("identity {s:?}: digest must be lowercase hex characters (0-9, a-f)"),
        ));
    }
    Ok((algorithm, digest))
}

/// Compute the canonical content Merkle-DAG identity of a source tree, returning
/// `"dag-sha256:<64-hex>"` (identity.md §1.8 / §2.1).
///
/// The on-disk tree is walked into a buffered `Vec<MaterializedEntry>` by
/// [`enumerate_local_entries`] (the canonical disk walk) and folded by
/// [`crate::dag_identity::compute_dag_identity`]. The STEP-1 invariant (a
/// git-materialized on-disk tree hashed here equals the git object-store
/// enumeration) is what lets every transport, plus CAS verify, share this one
/// identity site.
pub fn compute_content_hash(root: &Path) -> Result<String, CoreError> {
    compute_dag_identity(&enumerate_local_entries(root)?)
}

/// The canonical on-disk walk (the local materialize seam): walk `root` into a
/// buffered `Vec<MaterializedEntry>` (spec §1.8.4) feeding the epoch-2 DAG builder.
///
/// Single source of truth for turning real on-disk bytes + POSIX mode bits into
/// the materialize seam. [`compute_content_hash`] uses it, and the local transport
/// is exactly this walk applied to a local source dir — the local sibling of the
/// object-store seams [`crate::fetchers::enumerate_git_entries`] /
/// [`crate::fetchers::enumerate_tarball_entries`]. Mode mapping (spec §1.8.2.1):
/// a regular file with any POSIX execute bit (`st_mode & 0o111`) → `0x01`, else
/// `0x00`; a symlink → `0x80` with the `read_link` target string bytes (not
/// followed). A `.git` component is skipped at any depth (§1.8.6). Non-UTF-8
/// names / symlink targets are coded errors (cross-impl contract with Python).
pub fn enumerate_local_entries(root: &Path) -> Result<Vec<MaterializedEntry>, CoreError> {
    let mut out = Vec::new();
    walk(root, root, &mut out)?;
    Ok(out)
}

fn walk(root: &Path, dir: &Path, out: &mut Vec<MaterializedEntry>) -> Result<(), CoreError> {
    use std::os::unix::fs::PermissionsExt;

    let mut children: Vec<_> = std::fs::read_dir(dir)
        .map_err(|e| {
            CoreError::Identity(
                INTERNAL_IO,
                format!("cannot read directory {}: {e}", dir.display()),
            )
        })?
        .collect::<Result<Vec<_>, _>>()
        .map_err(|e| {
            CoreError::Identity(
                INTERNAL_IO,
                format!("cannot read entry under {}: {e}", dir.display()),
            )
        })?;
    // Stream order is irrelevant (the builder re-sorts each level by leaf name);
    // a stable sort keeps the walk deterministic.
    children.sort_by_key(|e| e.file_name());

    for ent in children {
        let path = ent.path();
        // §1.8.6 (inherits §1.4): exclude `.git` (and anything beneath it) at any depth.
        if path.file_name().map(|n| n == ".git").unwrap_or(false) {
            continue;
        }
        // §1.3 / spec/errors.md ID-NON-UTF8-RELPATH: the relpath is encoded as
        // UTF-8 (cross-impl contract with Python). On Linux, filenames are raw
        // byte sequences; validate before constructing the entry so both impls
        // raise the same coded error rather than silently diverging.
        let rb = relpath_bytes(root, &path);
        let relpath = match std::str::from_utf8(&rb) {
            Ok(s) => s.to_string(),
            Err(_) => {
                return Err(CoreError::Identity(
                    "ID-NON-UTF8-RELPATH",
                    format!(
                        "file path {:?} is not valid UTF-8 — cannot compute a content hash",
                        String::from_utf8_lossy(&rb)
                    ),
                ))
            }
        };
        // symlink_metadata does NOT follow the link, so a symlink — even to a
        // directory — reports as a symlink and becomes a leaf entry.
        let meta = std::fs::symlink_metadata(&path).map_err(|e| {
            CoreError::Identity(INTERNAL_IO, format!("cannot stat {}: {e}", path.display()))
        })?;
        let ft = meta.file_type();
        if ft.is_symlink() {
            // §1.5/§1.8.1: hash the link-target string as UTF-8 bytes; do not follow.
            let target = std::fs::read_link(&path).map_err(|e| {
                CoreError::Identity(
                    INTERNAL_IO,
                    format!("cannot readlink {}: {e}", path.display()),
                )
            })?;
            let target_str = target.into_os_string().into_string().map_err(|_| {
                CoreError::Identity(
                    "ID-NON-UTF8-SYMLINK-TARGET",
                    format!(
                        "symlink target at {relpath:?} is not valid UTF-8 \
                         — cannot compute a content hash"
                    ),
                )
            })?;
            out.push(MaterializedEntry {
                relpath,
                mode_byte: MODE_SYMLINK,
                content: target_str.into_bytes(),
            });
        } else if ft.is_dir() {
            walk(root, &path, out)?;
        } else if ft.is_file() {
            let content = std::fs::read(&path).map_err(|e| {
                CoreError::Identity(
                    INTERNAL_IO,
                    format!("cannot read file {}: {e}", path.display()),
                )
            })?;
            // §1.8.2.1: the executable bit is part of epoch-2 identity.
            let mode_byte = if meta.permissions().mode() & 0o111 != 0 {
                MODE_EXECUTABLE
            } else {
                MODE_REGULAR
            };
            out.push(MaterializedEntry {
                relpath,
                mode_byte,
                content,
            });
        }
        // else: fifo/socket/device — not representable; skipped, as Python's
        // is_file()/is_symlink() pair would also skip them.
    }
    Ok(())
}

/// The POSIX relative path from `root` to `path`, as raw bytes. On Linux a
/// filename's OS bytes equal its UTF-8 encoding when valid; using the raw bytes
/// both matches the spec's UTF-8 clause for valid paths and gives the exact
/// byte-order the §1.3 sort requires, with no lossy conversion.
fn relpath_bytes(root: &Path, path: &Path) -> Vec<u8> {
    let rel = path.strip_prefix(root).unwrap_or(path);
    rel.as_os_str().as_bytes().to_vec()
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::fs;
    use std::os::unix::fs::{symlink, PermissionsExt};

    // Owner-execute bit — used only by test helper `write` to create fixture
    // files with the exec bit set; the bit itself is no longer part of identity
    // (Resolved Decision 1).
    const S_IXUSR: u32 = 0o100;

    #[test]
    fn dag_sha256_is_supported_with_64_hex_chars() {
        assert_eq!(
            SUPPORTED_ALGORITHMS.iter().find(|(a, _)| *a == "dag-sha256"),
            Some(&("dag-sha256", 64))
        );
    }

    #[test]
    fn sha256_is_not_in_supported_algorithms() {
        assert!(
            SUPPORTED_ALGORITHMS.iter().all(|(a, _)| *a != "sha256"),
            "stale sha256 must not be in SUPPORTED_ALGORITHMS"
        );
    }

    // --- parse_identity (mirrors test_identity_validator.py) -----------------

    #[test]
    fn parse_identity_accepts_dag_sha256_canonical_form() {
        let valid = format!("dag-sha256:{}", "a".repeat(64));
        assert_eq!(
            parse_identity(&valid).unwrap(),
            ("dag-sha256", &"a".repeat(64)[..])
        );
    }

    #[test]
    fn parse_identity_rejects_stale_sha256_prefix() {
        let stale = format!("sha256:{}", "a".repeat(64));
        let e = parse_identity(&stale).unwrap_err();
        assert_eq!(e.code(), "ID-UNSUPPORTED-ALGORITHM");
    }

    #[test]
    fn parse_identity_rejects_unknown_algorithm() {
        let e = parse_identity(&format!("md5:{}", "a".repeat(32))).unwrap_err();
        assert_eq!(e.code(), "ID-UNSUPPORTED-ALGORITHM");
    }

    #[test]
    fn parse_identity_rejects_bare_hex_no_prefix() {
        let e = parse_identity(&"a".repeat(64)).unwrap_err();
        assert_eq!(e.code(), "ID-NO-ALGORITHM-PREFIX");
    }

    #[test]
    fn parse_identity_rejects_wrong_length_digest() {
        let e = parse_identity("dag-sha256:abc123").unwrap_err();
        assert_eq!(e.code(), "ID-WRONG-DIGEST-LENGTH");
    }

    #[test]
    fn parse_identity_rejects_non_hex_characters_in_digest() {
        let e = parse_identity(&format!("dag-sha256:{}", "Z".repeat(64))).unwrap_err();
        assert_eq!(e.code(), "ID-NON-HEX-DIGEST");
    }

    // --- compute_content_hash (mirrors test_identity.py) ---------------------

    fn write(path: &Path, content: &str, executable: bool) {
        fs::create_dir_all(path.parent().unwrap()).unwrap();
        fs::write(path, content).unwrap();
        if executable {
            let mut p = fs::metadata(path).unwrap().permissions();
            p.set_mode(p.mode() | S_IXUSR);
            fs::set_permissions(path, p).unwrap();
        }
    }

    fn tmp() -> std::path::PathBuf {
        // Unique scratch under the cargo target tmp; index-free via a counter.
        use std::sync::atomic::{AtomicU64, Ordering};
        static N: AtomicU64 = AtomicU64::new(0);
        let n = N.fetch_add(1, Ordering::Relaxed);
        let dir = std::env::temp_dir().join(format!("milpa-id-test-{}-{n}", std::process::id()));
        fs::create_dir_all(&dir).unwrap();
        dir
    }

    #[test]
    fn executable_bit_changes_content_hash() {
        // Epoch 2 (§1.8.2): the exec bit IS part of identity. A regular file
        // (0x00) and an executable file (0x01) with identical bytes hash
        // differently — a deliberate correction over interim epoch 1.
        let root = tmp();
        let (a, b) = (root.join("a"), root.join("b"));
        write(&a.join("script.sh"), "#!/bin/sh\necho hi\n", false);
        write(&b.join("script.sh"), "#!/bin/sh\necho hi\n", true);
        assert_ne!(
            compute_content_hash(&a).unwrap(),
            compute_content_hash(&b).unwrap()
        );
    }

    #[test]
    fn identical_trees_produce_identical_hashes() {
        let root = tmp();
        let (a, b) = (root.join("a"), root.join("b"));
        let files = [
            (
                "src/main.nim",
                "import std/strutils\nproc main() = echo \"hi\"\n",
                false,
            ),
            ("README.md", "# project\n", false),
            (
                "scripts/run.sh",
                "#!/bin/sh\nexec nim r src/main.nim\n",
                true,
            ),
        ];
        for (rel, content, exec) in files {
            write(&a.join(rel), content, exec);
            write(&b.join(rel), content, exec);
        }
        assert_eq!(
            compute_content_hash(&a).unwrap(),
            compute_content_hash(&b).unwrap()
        );
    }

    #[test]
    fn dot_git_directory_excluded() {
        let root = tmp();
        let (a, b) = (root.join("a"), root.join("b"));
        write(&a.join("src/main.nim"), "echo 'hi'\n", false);
        write(&b.join("src/main.nim"), "echo 'hi'\n", false);
        write(&b.join(".git/HEAD"), "ref: refs/heads/main\n", false);
        write(&b.join(".git/objects/ab/cd1234"), "binary junk", false);
        assert_eq!(
            compute_content_hash(&a).unwrap(),
            compute_content_hash(&b).unwrap()
        );
    }

    #[test]
    fn symlink_hashed_by_target_not_followed() {
        let root = tmp();
        let (a, b) = (root.join("a"), root.join("b"));
        fs::create_dir_all(&a).unwrap();
        fs::create_dir_all(&b).unwrap();
        fs::write(a.join("real.txt"), "hello\n").unwrap();
        symlink("real.txt", a.join("link")).unwrap();
        fs::write(b.join("real.txt"), "hello\n").unwrap();
        symlink("different_target", b.join("link")).unwrap();
        assert_ne!(
            compute_content_hash(&a).unwrap(),
            compute_content_hash(&b).unwrap()
        );
    }

    #[test]
    fn symlink_vs_regular_file_with_same_content_differ() {
        let root = tmp();
        let (a, b) = (root.join("a"), root.join("b"));
        fs::create_dir_all(&a).unwrap();
        fs::create_dir_all(&b).unwrap();
        fs::write(a.join("entry"), "target").unwrap();
        symlink("target", b.join("entry")).unwrap();
        assert_ne!(
            compute_content_hash(&a).unwrap(),
            compute_content_hash(&b).unwrap()
        );
    }

    #[test]
    fn symlink_pointing_outside_tree_does_not_crash() {
        let root = tmp();
        let a = root.join("a");
        fs::create_dir_all(&a).unwrap();
        symlink("/nonexistent/elsewhere", a.join("broken")).unwrap();
        let h = compute_content_hash(&a).unwrap();
        let digest = h.strip_prefix("dag-sha256:").unwrap();
        assert_eq!(digest.len(), 64);
        assert!(digest
            .bytes()
            .all(|b| b.is_ascii_hexdigit() && !b.is_ascii_uppercase()));
    }

    #[test]
    fn byte_parity_with_python_oracle() {
        // The digest below is produced by the Python reference impl
        // (milpa.identity.compute_content_hash) over a tree with one regular
        // file, one executable file, a symlink, and an excluded `.git/`. If the
        // Rust byte stream ever drifts from the spec, this breaks. Regenerate
        // ONLY by re-running the oracle on the identical tree.
        let a = tmp().join("t");
        write(
            &a.join("src/main.nim"),
            "import std/strutils\nproc main() = echo \"hi\"\n",
            false,
        );
        write(&a.join("README.md"), "# project\n", false);
        write(
            &a.join("run.sh"),
            "#!/bin/sh\nexec nim r src/main.nim\n",
            true,
        );
        symlink("src/main.nim", a.join("mainlink")).unwrap();
        write(&a.join(".git/HEAD"), "junk\n", false);
        assert_eq!(
            compute_content_hash(&a).unwrap(),
            "dag-sha256:10c68c24594e4ab384f0672a67b738eab5190e0837a3e09dd27e89eb1172791a"
        );
    }

    #[test]
    fn empty_tree_hashes_the_empty_stream() {
        let root = tmp();
        let a = root.join("a");
        fs::create_dir_all(&a).unwrap();
        // Empty tree = the zero-entry Merkle-DAG root (§1.8.5): sha256(b"").
        assert_eq!(
            compute_content_hash(&a).unwrap(),
            "dag-sha256:e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"
        );
    }

    // --- ID-NON-UTF8-RELPATH (spec/errors.md; distinct from
    // --- ID-NON-UTF8-SYMLINK-TARGET which covers symlink *targets*)

    #[test]
    fn non_utf8_relpath_in_regular_file_raises_coded_error() {
        // A source tree containing a file whose *name* (not content) contains
        // non-UTF-8 bytes must raise ID-NON-UTF8-RELPATH, not panic or silently
        // produce a wrong hash.  Mirrors the Python test and the
        // ID-NON-UTF8-SYMLINK-TARGET test pattern.
        //
        // This test creates a file via raw OS bytes (std::fs + OsStr::from_bytes).
        // Skip gracefully if the filesystem rejects the bytes (vfat, some WSL
        // mounts), analogous to the Python "pytest.skip" branch.
        use std::ffi::OsStr;
        use std::os::unix::ffi::OsStrExt as _;

        let root = tmp();
        let a = root.join("a");
        fs::create_dir_all(&a).unwrap();

        // Build a file path with a raw non-UTF-8 byte (0xff) in the name.
        let bad_name = OsStr::from_bytes(b"\xff\xfe");
        let bad_path = a.join(bad_name);
        match fs::write(&bad_path, b"content") {
            Err(_) => {
                // Filesystem rejected the non-UTF-8 name — skip gracefully.
                return;
            }
            Ok(()) => {}
        }

        let err = compute_content_hash(&a).unwrap_err();
        assert_eq!(
            err.code(),
            "ID-NON-UTF8-RELPATH",
            "expected ID-NON-UTF8-RELPATH, got {:?}",
            err
        );
    }

    #[test]
    fn non_utf8_relpath_in_symlink_name_raises_coded_error() {
        // A symlink whose *name* (not target) contains non-UTF-8 bytes must also
        // raise ID-NON-UTF8-RELPATH.  The relpath check fires before the
        // symlink-target check, so the symlink-target itself is irrelevant here.
        use std::ffi::OsStr;
        use std::os::unix::ffi::OsStrExt as _;

        let root = tmp();
        let a = root.join("a");
        fs::create_dir_all(&a).unwrap();

        let bad_name = OsStr::from_bytes(b"\xff\xfe");
        let bad_link_path = a.join(bad_name);
        match symlink("some_target", &bad_link_path) {
            Err(_) => {
                // Filesystem rejected the non-UTF-8 name — skip gracefully.
                return;
            }
            Ok(()) => {}
        }

        let err = compute_content_hash(&a).unwrap_err();
        assert_eq!(
            err.code(),
            "ID-NON-UTF8-RELPATH",
            "expected ID-NON-UTF8-RELPATH, got {:?}",
            err
        );
    }

    #[test]
    fn non_utf8_symlink_target_still_raises_symlink_target_code() {
        // Sanity: a symlink with a VALID UTF-8 name but a NON-UTF-8 target
        // still raises ID-NON-UTF8-SYMLINK-TARGET (not ID-NON-UTF8-RELPATH),
        // confirming the two codes are distinct and the right one fires first.
        let root = tmp();
        let a = root.join("a");
        fs::create_dir_all(&a).unwrap();

        use std::ffi::OsStr;
        use std::os::unix::ffi::OsStrExt as _;
        let bad_target = OsStr::from_bytes(b"\xff\xfe");
        // valid UTF-8 link name, non-UTF-8 target
        match std::os::unix::fs::symlink(bad_target, a.join("valid_name_link")) {
            Err(_) => return, // filesystem rejected the target bytes — skip
            Ok(()) => {}
        }

        let err = compute_content_hash(&a).unwrap_err();
        assert_eq!(
            err.code(),
            "ID-NON-UTF8-SYMLINK-TARGET",
            "expected ID-NON-UTF8-SYMLINK-TARGET for non-UTF-8 target, got {:?}",
            err
        );
    }
}
